lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
+
+  // To finish up the conversion, remove all the references to the old type by
+  // carefully replacing them with references to the new one.
+  return replaceAllUsesWith(Function.key(), NewTypePath, Binary);
+}
+
+using TDManager = TypeDependencyManager;
+template<typename RT>
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingRegisterArguments(const SortedVector<RT> &Used) {
+  llvm::SmallVector<model::Register::Values, 8> ArgumentRegisters;
+
+  // Rely onto the register state deduction to make sure no "holes" are
+  // present in-between the argument registers.
+  abi::RegisterState::Map Map(model::ABI::getArchitecture(ABI.ABI()));
+  for (const RT &Reg : Used)
+    Map[Reg.Location()].IsUsedForPassingArguments = abi::RegisterState::Yes;
+  auto Enforced = abi::enforceRegisterStateDeductions(Map, ABI);
+  for (auto [Register, Pair] : Enforced)
+    if (abi::RegisterState::shouldEmit(Pair.IsUsedForPassingArguments))
+      ArgumentRegisters.emplace_back(Register);
+
+  // But just knowing which registers we need is not sufficient, we also have
+  // to order them properly.
+  auto Ordered = abi::orderArguments(ArgumentRegisters, ABI);
+
+  llvm::SmallVector<model::Argument, 8> Result;
+  for (model::Register::Values Register : Ordered) {
+    auto CurrentRegisterIterator = Used.find(Register);
+    if (CurrentRegisterIterator != Used.end()) {
+      // If the current register is confirmed to be in use, convert it into
+      // an argument while preserving its type and name.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    } else {
+      // This register is unused but we still have to create an argument
+      // for it. Otherwise the resulting function will not be compatible with
+      // the current one: if the "hole" is not preserved, there will be no way
+      // to recreate it at any point in the future, when, for example,
+      // converting back to the original-like representation.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
```

Comment:
You're not copying over `OriginalName`. Please check if there are other interesting fields.

Reply:
Whoops, this code is older than `OriginalName`, I think.

When `refactoring` it I was just mindlessly updating the fields without thinking about the new ones. My bad.

Reply:
Maybe we could have a `copyNames` primitive in `revng/Model/Helpers.h`.
This typical and it will get worse in the future: on top of `CustomName` and `OriginalName` we'll also have `Comment`.
Maybe `copyMetadata`?

Reply:
There is no `renvg/Model/Helpers.h` 😅. I placed it into `revng/Model/Processing.h` for now: https://github.com/revng/revng/pull/232/commits/8c7f1f83af4c7f962994c888589c3e3511993fba.

What do you think?

Reply:
No biggie, but that file is more for things the make "global" changes to the model.
Consider creating `revng/Model/Helpers.h` or `revng/Model/Metadata.h`?

Reply:
Sounds good!

Reply:
It's in `revng/Model/Helpers.h` now: https://github.com/revng/revng/pull/232/commits/717f5a9bee41a4bad64c2ae97b310ee25ab251f6

Reply:
Closing this.